Ot 26 endpoint de autenticacion

devmontini opened the pull request.

devmontini committed:
Se instalo el paquete de express validator
Se agrego al modelo de User id el "primaryKey"
Se agrego la ruta "/auth" al archivo app.js
Se creo la estructura de la ruta de autentificacion "/login"
Se creo estructura base para el metodo login de controlador "autentificacion"

devmontini committed:
Auth routes se creo la validacion de datos login
Controller aut se valida si el email existe

devmontini committed:
Se instalo la dependencia bcryptjs para encriptar
Se agrego logica de comparacion de pass en controllers

devmontini committed:
Se agrego la respuesta { ok: false } al controller

soyarielruiz commented on controllers/authentication.js:
```diff
@@ -0,0 +1,26 @@
+
+class AuthController {
+  async loginUser(email, pass) {
+    //Primero validamos el email
```
nit: comentario que no suma información.

soyarielruiz commented on controllers/authentication.js:
```diff
@@ -0,0 +1,26 @@
+class AuthController {
+  async loginUser(email, pass) {
+    //Primero validamos el email
+    const userInfo = await User.findAll({
```
En caso que haya un error en la llamada a la DB no se está atrapando el error.

soyarielruiz commented on controllers/authentication.js:
```diff
@@ -0,0 +1,26 @@
+      where: {
+        email: email,
+      },
+    });
```
Deberías definir que campos queres devolver de la DB así no expones todos los campos.

soyarielruiz commented on routes/authentication.js:
```diff
@@ -0,0 +1,32 @@
+
+const AuthController = require('../controllers/authentication.js');
+
+const authController = new AuthController();
```
Esto genera una reserva de memoria que se ejecuta con el servidor aún cuando no se ha ejecutado, revisar.

devmontini replied:
Lo pongo de esta manera
`
const AuthController = require('../controllers/authentication.js');

router.post(
  ...
      const response = await AuthController.loginUser(email, pass);
  ...
);
`

Pero me sale el error: "TypeError: AuthController.loginUser is not a function"

soyarielruiz commented on routes/authentication.js:
```diff
@@ -0,0 +1,32 @@
+  body('email').isEmail(),
+  body('password').isLength({ min: 4 }),
+  async (req, res, next) => {
+    try {
```
Esta validación se puede generalizar para que pueda usarse fuera del router, sino se repetirá código.

soyarielruiz replied:
El router no tiene responsabilidades de validar flujos.

devmontini committed:
Solucionado controllers comentarios

devmontini committed:
Se arreglo la validacio en las rutas
Se borraron comentarios
Se definieron los campos que traemos de la DB

devmontini committed:
Se modifico el controler en modo static
Y se soluciono el problema de uso de memoria en la ruta auth

devmontini commented:
Listo los arreglos

soyarielruiz commented on controllers/authentication.js:
```diff
@@ -0,0 +1,30 @@
+
+class AuthController {
+  static async loginUser(email, pass) {
+    try {
```
nit: Bloque try debe ser puntual para manejo de errores.

soyarielruiz commented on controllers/authentication.js:
```diff
@@ -0,0 +1,30 @@
+
+      return userInfo;
+    } catch (error) {
+      console.log(error);
```
nit:_console.log_ no es funcional para saber que pasó con el flujo, debería lanzar una excepción/error.

soyarielruiz commented on routes/authentication.js:
```diff
@@ -0,0 +1,18 @@
+
+router.post('/login', loginValidation, async (req, res, next) => {
+  try {
+    const email = req.body.email;
```
La route no debe encargarse del manejo de la respuesta.

devmontini committed:
Sacamos la logica de la ruta routes/authentication
Corregimos el manejo de errores en controlelrs/authentication
Movimos el llamado a la base de datos a dao/authentication
Editamos el nombre del archivo validator y creamos el validator(login)
Creamos el schema para usar en la ruta y el validator
Agregamos db:reset en package.json
Modifique el seeders de user para agregar usuario con pass encryptada

devmontini committed:
Merge remote-tracking branch 'origin/development' into OT-26_Endpoint_de_autenticacion

soyarielruiz commented on controllers/authentication.js:
```diff
@@ -0,0 +1,23 @@
+
+class AuthController {
+  static async loginUser(req, res) {
+    const { email } = req.body;
```
nit: se puede hacer en una misma línea.

soyarielruiz commented on dao/authentication.js:
```diff
@@ -0,0 +1,21 @@
+
+      return userData;
+    } catch (error) {
+      return res.status(400).json(error);
```
El http status solo se devuelva como response, no como manejo interno de tu flujo.

soyarielruiz commented on dao/authentication.js:
```diff
@@ -0,0 +1,21 @@
+const { User } = db;
+
+class AuthDao {
+  static async _findUser(email) {
```
nit: lo haría abstracto para que busque un usuario mediante un objeto clave valor para asociarlo al where y sea adaptable a cualquier otra columna que se quiera usar.

soyarielruiz commented on seeders/20201109013159-create-demo-user.js:
```diff
@@ -2,17 +2,37 @@
+    );
   },
 
   down: async (queryInterface, Sequelize) => {
```
Ya que se tocó este file sería bueno agregar el caso de rollback.

devmontini replied:
el demo user es de tomas y no se si hizo el mege del modelo user a development todavia

soyarielruiz commented on schemas/login.js:
```diff
@@ -0,0 +1,12 @@
+const { check } = require('express-validator');
```
nit: Habría que ver si definir un schema por router es lo conveniente.

devmontini replied:
Perfecto mañana lo comento con el equipo.

devmontini committed:
Merge branch 'development' into OT-26_Endpoint_de_autenticacion

devmontini committed:
Se agrego un seeder de user con password encriptada y se agrego el rollback

devmontini committed:
Merge branch 'FIX-Error_Activities_New_Testimonial_Packages' into OT-26_Endpoint_de_autenticacion

devmontini committed:
Se mejoro:
- La logica y la respuesta del controller de authentication
- La funcion _findUser de Dao ahora es abtracta y adaptable
- Se agrego el rollback a create demo user

soyarielruiz commented on dao/authentication.js:
```diff
@@ -0,0 +1,23 @@
+const { User } = db;
+class AuthDao {
+  /**
+   * Asynchronously return a user from the database (table Users)
```
👏

soyarielruiz commented on dao/authentication.js:
```diff
@@ -0,0 +1,23 @@
+   * @param {Object} objKeyValues - To filter by attribute/s Ej: {email: 'user@mail'}
+   * @param {string[]} arrayAttributes - Attributes we need Ej: ['username', 'age']
+   */
+  static async _findUser(objKeyValues, arrayAttributes) {
```
nit: si es **_** al ppio es privado y es lo que no queres.

soyarielruiz commented on controllers/authentication.js:
```diff
@@ -0,0 +1,25 @@
+    const { email, password } = req.body;
+    try {
+      const objKeyValues = { email: email };
+      const arrayAttributes = ['email', 'password'];
```
nit: el nombre de la variable no debe indicar el tipo, eso es para otros lenguajes.

soyarielruiz commented on validators/login.js:
```diff
@@ -0,0 +1,13 @@
+const { validationResult } = require('express-validator');
+
+const validateResult = (req, res, next) => {
```
Código funcional pasar a POO y Maxi hizo algo similar.

devmontini committed:
Se actualizo rama local con development y se arreglaron los conflictos

devmontini committed:
SE creo y soluciono:
Se paso a POO el validator y se unio al archivo de maxi
Se puso el schema de login dentro de user
Se saco el _ del dao
Se cambiaron los nombres de variables

devmontini committed:
Merge branch 'development' into OT-26
Se solucionaron conflictos con los nuevos cambios

devmontini committed:
Se modifico el nombre de variables

soyarielruiz commented on middlewares/validationErrors.js:
```diff
@@ -14,6 +14,17 @@ class ValidationErrors {
     }
   }
+
+  static async validateLogin(req, res, next) {
```
Código duplicado, la función de la línea 4 hace lo mismo.

devmontini committed:
Merge branch 'development' into OT-26_Endpoint_de_autenticacion

devmontini committed:
Eliminada funcion duplicada de validacion

soyarielruiz reviewed (approved):
lgtm
